pallets/external-validator-slashes/src/lib.rs

```diff
@@ -0,0 +1,501 @@
+    parity_scale_codec::{Decode, Encode},
+    sp_runtime::traits::{Convert, Debug, One, Saturating, Zero},
+    sp_runtime::DispatchResult,
+    sp_runtime::Perbill,
+    sp_staking::{
+        offence::{OffenceDetails, OnOffenceHandler},
+        EraIndex, SessionIndex,
+    },
+    sp_std::vec,
+    sp_std::vec::Vec,
+    tp_traits::{EraIndexProvider, InvulnerablesProvider, OnEraStart},
+};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+pub mod weights;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    pub use crate::weights::WeightInfo;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Removed author data
+        SlashReported {
+            validator: T::ValidatorId,
+            fraction: Perbill,
+            slash_era: EraIndex,
+        },
+    }
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// A stable ID for a validator.
+        type ValidatorId: Member
+            + Parameter
+            + MaybeSerializeDeserialize
+            + MaxEncodedLen
+            + TryFrom<Self::AccountId>;
+
+        /// A conversion from account ID to validator ID.
+        type ValidatorIdOf: Convert<Self::AccountId, Option<Self::ValidatorId>>;
+
+        /// Number of eras that slashes are deferred by, after computation.
+        ///
+        /// This should be less than the bonding duration. Set to 0 if slashes
+        /// should be applied immediately, without opportunity for intervention.
+        #[pallet::constant]
+        type SlashDeferDuration: Get<EraIndex>;
+
+        /// Number of eras that staked funds must remain bonded for.
+        #[pallet::constant]
+        type BondingDuration: Get<EraIndex>;
+
+        // SlashId type, used as a counter on the number of slashes
+        type SlashId: Default
+            + FullCodec
+            + TypeInfo
+            + Copy
+            + Clone
+            + Debug
+            + Eq
+            + Saturating
+            + One
+            + Ord
+            + MaxEncodedLen;
+
+        /// Interface for interacting with a session pallet.
+        type SessionInterface: SessionInterface<Self::AccountId>;
+
+        /// Era index provider, used to fetch the active era among other thins
+        type EraIndexProvider: EraIndexProvider;
+
+        /// Invulnerable provider, used to get the invulnerables to know when not to slash
+        type InvulnerablesProvider: InvulnerablesProvider<Self::ValidatorId>;
+
+        /// The weight information of this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        EmptyTargets,
+        InvalidSlashIndex,
+        NotSortedAndUnique,
```

Comment:
Would be nice to have at least one short line of explanation docs for each error.

Reply:
fixed